src/jdk.incubator.concurrent/share/classes/jdk/incubator/concurrent/ScopedValue.java

```diff
@@ -0,0 +1,837 @@
+ * scoped values that it uses cyclically, the cache hit rate
+ * will be low and performance will be poor. This design allows
+ * scoped-value inheritance by {@link StructuredTaskScope} threads to
+ * be very fast: in essence, no more than copying a pointer, and
+ * leaving a scoped-value binding also requires little more than
+ * updating a pointer.
+ *
+ * <p>Because the scoped-value per-thread cache is small, clients
+ * should minimize the number of bound scoped values in use. For
+ * example, if it is necessary to pass a number of values in this way,
+ * it makes sense to create a record class to hold those values, and
+ * then bind a single {@code ScopedValue} to an instance of that record.
+ *
+ * <p>For this incubator release, the reference implementation
+ * provides some system properties to tune the performance of scoped
+ * values.
+ *
+ * <p>The system property {@code jdk.incubator.concurrent.ScopedValue.cacheSize}
+ * controls the size of the (per-thread) scoped-value cache. This cache is crucial
+ * for the performance of scoped values. If it is too small,
+ * the runtime library will repeatedly need to scan for each
+ * {@link #get}. If it is too large, memory will be unnecessarily
+ * consumed. The default scoped-value cache size is 16 entries. It may
+ * be varied from 2 to 16 entries in size. {@code ScopedValue.cacheSize}
+ * must be an integer power of 2.
+ *
+ * <p>For example, you could use {@code -Djdk.incubator.concurrent.ScopedValue.cacheSize=8}.
+ *
+ * <p>The other system property is {@code jdk.preserveScopedValueCache}.
+ * This property determines whether the per-thread scoped-value
+ * cache is preserved when a virtual thread is blocked. By default
+ * this property is set to {@code true}, meaning that every virtual
+ * thread preserves its scoped-value cache when blocked. Like {@code
+ * ScopedValue.cacheSize}, this is a space versus speed trade-off: in
+ * situations where many virtual threads are blocked most of the time,
+ * setting this property to {@code false} might result in a useful
+ * memory saving, but each virtual thread's scoped-value cache would
+ * have to be regenerated after a blocking operation.
+ *
+ * @param <T> the type of the object bound to this {@code ScopedValue}
+ * @since 20
+ */
+public final class ScopedValue<T> {
+    private static final JavaLangAccess JLA = SharedSecrets.getJavaLangAccess();
+
+    private final @Stable int hash;
+
+    @Override
+    public int hashCode() { return hash; }
+
+    /**
+     * An immutable map from {@code ScopedValue} to values.
+     *
+     * <p> Unless otherwise specified, passing a {@code null} argument to a constructor
+     * or method in this class will cause a {@link NullPointerException} to be thrown.
+     */
+    static final class Snapshot {
+        final Snapshot prev;
+        final Carrier bindings;
+        final int bitmask;
+
+        private static final Object NIL = new Object();
```

Comment:
I agree its not very important, please feel free to ignore my suggestion. My thinking was to prefer something more explicit in the code when reading, since i felt the use of the term nil was more idiomatic in other languages than Java.